features/430-passwordless/config.md

```diff
@@ -1 +1,71 @@
-TODO
\ No newline at end of file
+# Configuration
+
+Configurations are refactored to better match the current identity model:
+
+## Change Summary
+- Email sender & reply-to become a global option, applying to all email
+  messages.
+- Email subjects can be localized like OIDC client metadata.
+- Identity conflict policy is grouped together as a configuration group.
+- MFA config are generalized to authenticator configs.
+- Identity type configs are grouped together.
+
+# Sample Configuration
+```yaml
+smtp:
+    sender: 'admin@example.com'
+    reply_to: 'support@example.com'
+
+auth_api:
+    enabled: true
+    on_identity_conflict:
+        login_id:
+            allow_create_new_user: false
+        oauth:
+            allow_create_new_user: true
+            allow_auto_merge_user: false
+
+authentication:
+    identities: [login_id, oauth, anonymous]
+    primary_authenticators: [password, oauth]
+    secondary_authenticators: [totp, oob, bearer_token]
+
+authenticator:
+    password:
+        policy:
+            min_length: 8
+    totp:
+        maximum: 1
+    oob:
+        sms:
+            maximum: 0
+            subject: SMS one-time password
+            subject#zh: SMS 一次性密碼
+        email:
+            maximum: 1
+            subject: Email one-time password
+    bearer_token: {} # ...
+    recovery_code: {} # ...
+
+identity:
+    login_id:
+        types: {} # ...
+        keys: {} # ...
+    oauth:
+        providers: {} # ...
+```
+
+# Authentication configuration
+The authentication configuration is in `authentication` key. It describes
+the authentication flow in Auth UI.
+
+- `identities`: string list, must be non-empty.
+                List of usable identity types.
+- `primary_authenticators`: string list, must be non-empty.
+                            List of primary authenticator types.
+- `secondary_authenticators`: string list, can be non-empty.
+                              List of secondary authenticator types.
+
+Identity and authenticator types must be included in one of the lists to be
+enabled. There must be no duplicated authenticators across primary and
```

Comment:
Do we need to define the authentication factors?

Suppose we agree that there are 3 authentication factors

- something the user has (A)
- something the user knows (B)
- something the user is (C)

The next step is to classify our authenticators into one of these factors

- oob: A
- totp: A
- bearer_token: A
- recovery_code: A
- anonymous: A
- webauthn: A
- password: B
- biometric: C
- oauth: C

The next question is should we enforce the two authenticators must be of different factors? Or we allow authenticators of the same factor as long as the authenticators are distinct?

If we allow the latter case then the configuration should allow duplicated authenticator types.

Reply:
I don't think we need to enforce this; the three factor kinds are not defined strictly (e.g. a password can be considered something user knows, but a written down password can be considered something user has), and mostly used for explanatory purpose.